src/Telegram.Bot/TelegramBotClient.cs

```diff
@@ -104,54 +113,130 @@ public TelegramBotClient(string token, HttpClient? httpClient = null)
             var reqDataArgs = new ApiRequestEventArgs(request.MethodName, httpRequest.Content);
             MakingApiRequest?.Invoke(this, reqDataArgs);
 
-            HttpResponseMessage httpResponse;
+            HttpResponseMessage? httpResponse;
             try
             {
                 httpResponse = await _httpClient.SendAsync(httpRequest, cancellationToken)
                     .ConfigureAwait(false);
             }
-            catch (TaskCanceledException e)
+            catch (Exception exception)
             {
-                if (cancellationToken.IsCancellationRequested)
-                    throw;
+                throw new RequestException("Exception during making request", exception);
+            }
+
+            try
+            {
+                // required since user might be able to set new status code using following event arg
+                var actualResponseStatusCode = httpResponse.StatusCode;
+                ApiResponseReceived?.Invoke(this, new ApiResponseEventArgs(httpResponse, reqDataArgs));
+
+                if (actualResponseStatusCode != HttpStatusCode.OK)
+                {
+                    var failedApiResponse = await httpResponse
+                        .DeserializeContentAsync<FailedApiResponse>(
+                            actualResponseStatusCode
+                        );
+
+                    if (failedApiResponse is null)
+                    {
+                        var stringifiedResponse = await httpResponse.Content
+                            .ReadAsStringAsync()
+                            .ConfigureAwait(false);
+
+                        throw new RequestException(
+                            "Required properties not found in response.",
+                            actualResponseStatusCode,
+                            stringifiedResponse
+                        );
+                    }
+
+                    throw ExceptionParser.Parse(
+                        failedApiResponse.ErrorCode,
+                        failedApiResponse.Description,
+                        failedApiResponse.Parameters
+                    );
+                }
+
+                var successfulApiResponse = await httpResponse
+                    .DeserializeContentAsync<SuccessfulApiResponse<TResult>>(
+                        actualResponseStatusCode
+                    );
```

Comment:
Shouldn't ever take the async path here anyway, but `.ConfigureAwait(false);` for consistency